smith-chem-wisc/mzLib
# Deconvolution Refactoring

* Structure
Deconvolution is being restructured to follow the Strategy Design Pattern from Design Patterns: Elements of Reusable Object-Oriented Software. 

![image](https://user-images.githubusercontent.com/96196865/198063701-1712df51-0889-4203-bc61-42d025e7318d.png)


Context - Deconvoluter
   DeconvoluterExtensions allows for deconvolution of objects other than MsDataScan with additional required input
Strategy - Deconvolution Algorithm
Abstract Strategy - DeconvolutionAlgoritm
Concrete Strategy - ClassicDeconvolutionAlgorithm, ExampleNewDeconvolutionAlgorithm
Abstract Parameters - DeconvolutionParameters
Concrete Parameters - ClassicDeconvolutionParameters, ExampleNewDeconvolutionParameters

- Setting Up Strategy 
   - Construct Specific DeconvolutionParameters Object
      - All fields which must be initialized are found within the constructor
      - Fields that have the option to be left out, are those that can only be found within specific objects (e.g. Isolation range is tied to MsDataScan)
   - Construct Deconvoluter(DeconvolutionType, DeconvolutionParameters)
      - Automatically constructs the Deconvolution Algorithm of choice  

![image](https://user-images.githubusercontent.com/96196865/199080862-14dc70b0-266c-4fa0-9c0f-2c790ca1c328.png)
   

- Using Strategy
   - Call Deconvoluter.Deconvolute() to receive an IEnumerable<IsotopicEnvelope> from the deconvoluted spectra
   - Call MsDataScan.GetIsolatedChargesAndMasses(Deconvoluter) to recieve an IEnumerable<IsotopicEnvelope> from the deconvoluted spectra where all isotopes have a peak within the isolation range
 
![image](https://user-images.githubusercontent.com/96196865/199081587-ec4f68ca-371e-46b1-a28a-84f4456d6979.png)

- Testing
   - All test that were previously implemented are now implemented for the new structure 
      - As the algoritm is the exact same, just moved, the tests are identical

- Adding New Algorithms
   - Add new parameter class to MassSpectrometry/Deconvolution/Parameters
      - Inherit class from DeconvolutionParameters
         - In constructor, require all fields that are necessary to the algorithm
            - if these values need to come from the MsDataScan object, set the default of the constructor parameters to null 
   - Add new algorithm class to MassSpectrometry/Deconvolution/Algorithms 
      - Inherit class from DeconvolutionAlgorithm, this should force implementation of three methods
         - Deconvolute() --> The new deconvolution algorithm
         - Constructor should remain empty
   - In Deconvoluter
      -  Add new deconvolution algorithm type to the DeconvoltionTypes Enum
      - Add case to switch within ConstructDeconvolutionAlgorithm() to create the new Deconvolution Algorithm
      - Add DeconvolutionExtensions for any objects besides MsDataScan that you would like to deconvolute
         - Required parameters should be (this Deconvoluter, {New object to deconvolute}, fields not initialized in parameters as they are found within MsDataScan)
         
![image](https://user-images.githubusercontent.com/96196865/199081530-fc58d947-8542-41d1-ba4e-350feeeee5d8.png)
![image](https://user-images.githubusercontent.com/96196865/199081478-deb66b1e-94f4-41f0-a1a8-d5b6676cfa52.png)

- Implementing in MetaMorpheus
   - Search task can instantiate private Deconvoluter and DeconvolutionParameter objects and call the Deconvoluter.Deconvolute() method
